src/cipherstate.rs

```diff
@@ -45,22 +46,23 @@ impl CipherState {
         authtext: &[u8],
         ciphertext: &[u8],
         out: &mut [u8],
-    ) -> Result<usize, ()> {
+    ) -> Result<usize, Error> {
         if (ciphertext.len() < TAGLEN) || (out.len() < (ciphertext.len() - TAGLEN) || !self.has_key)
         {
-            return Err(());
+            return Err(StateProblem::MissingKeyMaterial.into());
```

Comment:
This isn't just missing key material, since this if statement has length checks in it too. If we want to remain consistent, this should be `Error::Decrypt` since that's how it was handled previously.

In reality, I'd say we should break this if statement into two and serve either this missing key material error or the general decrypt error.

Reply:
I split them up, but while I was in the process of it I got slightly carried away with the validation done in `cipherstate`. I unified the checks between `CipherState` and `StatelessCipherState` with a small trait, with the added bonus of de-duplicating the encryption validations too.

If you don't like it, I'll undo and just stick with splitting them.